### PR TITLE
Add a Hardware section in the documentation

### PR DESCRIPTION
**Pull request by a-corni**
- Neutral-atom Hardware (A)
    - What are the components of a neutral-atom QC ? How does it work ?
    - Device
        - What are the physical constraints explicited in a Device ?
    - Channel
        - What are the components of a Channel ?
    - What are the example of devices present in Pulser ? (AnalogDevice, DigitalAnalogDevice, MockDevice)
        - The simplest example: AnalogDevice
            - Description of the contents of AnalogDevice

### CODE REVIEW

File: docs/source/hardware.ipynb [hunk 1]
```diff
@@ -178,7 +180,7 @@
    "metadata": {},
    "source": [
     ":::{note}\n",
-    "The modulation bandwidth and the EOM support are associated with more advanced features explained [in this tutorial](./tutorials/output_mod_eom.nblink). For a full list of the parameters of a `Channel` and the constraints they set, please check the [API documentation](./apidoc/core.rst#base-channels).\n",
+    "The modulation bandwidth and the EOM support are associated with more advanced features explained [in this tutorial](./tutorials/output_mod_eom.nblink). For a full list of the parameters of a `Channel` and the constraints they set, please check the [API documentation](./apidoc/core.html#base-channels).\n",
```
Comment (HGSilveri) [inline docs/source/hardware.ipynb:183]: This should work
```suggestion
    "The modulation bandwidth and the EOM support are associated with more advanced features explained [in this tutorial](./tutorials/output_mod_eom.nblink). For a full list of the parameters of a `Channel` and the constraints they set, please check the [API documentation](apidoc/core.rst#channels).\n",
```

**Review by HGSilveri — COMMENTED** [review]
Let's try to fix the build to help the review

File: docs/source/hardware.md [hunk 2]
```diff
@@ -0,0 +1,62 @@
+# Neutral-atom Hardware
+
+## How does a neutral-atom Quantum Computer work ?
```
[the remark this answers is not in the document]
Comment (HGSilveri) [reply docs/source/hardware.md]: Ah shit, these were old comments. Please ignore them

File: docs/source/hardware.ipynb (hunk 1 above)
Comment (a-corni) [reply docs/source/hardware.ipynb:183]: Ah sorry, I was not ready for review yet, it's still a draft :)

Comment (HGSilveri) [reply docs/source/hardware.ipynb]: Ah ok, I'll wait then :)

File: docs/source/hardware.ipynb [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+   "metadata": {},
+   "source": [
+    "## The `Device`"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "As presented in the [introduction to programming a neutral-atom QPU](programming.md#writing-a-pulser-program), the first step to writing a Pulser program is [the selection of a `Device`](programming.md#pick-a-device). \n",
```
Comment (ferrulli1pasqal) [inline docs/source/hardware.ipynb:30]: issue in rendering

File: docs/source/hardware.ipynb [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+    "\n",
+    "The `Device` object stores **all the physical constraints a quantum program written with `Pulser` should verify**, for it to be run on a neutral-atom QPU. The quantum programs, the `pulser.Sequence`, verify that each operation added to them are valid with respect to the constraints of the `Device`.  "
```
Comment (ferrulli1pasqal) [inline docs/source/hardware.ipynb:32]: I just realised that we never mentioned in the quick-start notebooks about the sequence initialisation besides a short sentence. Probably worth a step itself.

File: docs/source/hardware.ipynb [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+    "The `Device` represents the physics of a neutral-atom QPU but is not the QPU itself. A QPU must be accessed via a `Backend`, which is presented [in this notebook](./apidoc/backend.rst).\n",
+    "\n",
+    "</div>"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "If you want to run a quantum program on a QPU, start by selecting the `Backend` associated with this QPU. From this `Backend` the `Device` associated with the QPU can be accessed via the `get_available_devices` method.\n",
```
Comment (ferrulli1pasqal) [inline docs/source/hardware.ipynb:60]: I feel we're mixing up the goals: do we want to present the devices or something different? I feel confused reading this sentence. Same applies for next two paragraphs.

File: docs/source/hardware.ipynb (hunk 4 above)
[the remark this answers is not in the document]
Comment (a-corni) [reply docs/source/hardware.ipynb]: - **When it comes to running a program on a QPU**: Each QPU has an associated `Device`, which you can [get from the cloud provider you use to access this QPU](tutorials/backends.nblink#1.2.-Preparation-for-execution-on-QPUBackend).

- **For sole designing purposes**: `Pulser` provides examples of typical physical devices in `pulser.devices`. Notably, `pulser.AnalogDevice` is an example of a QPU implementing an [Ising Hamiltonian](./programming.md#ising-hamiltonian). 

- **For designing without physical constraints** during the writing of your quantum program, you can use a `VirtualDevice`. An example of such a `VirtualDevice` is the `MockDevice` provided in the `pulser.devices`, which gives full liberty to write a quantum program. `VirtualDevice` is detailed in [an advanced tutorial](./tutorials/virtual_devices.nblink).

File: docs/source/hardware.ipynb [hunk 6]
```diff
@@ -0,0 +1,247 @@
...
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "<div class=\"alert alert-info\">\n",
+    "\n",
+    "**Important note**:\n",
+    "\n",
+    "\n",
+    "The `Device` represents the physics of a neutral-atom QPU but is not the QPU itself. A QPU must be accessed via a `Backend`, which is presented [in this section](./tutorials/backends.nblink).\n",
```
Comment (ferrulli1pasqal) [inline docs/source/hardware.ipynb:46]: If I understand what is the message you want to convey, I'd rephrase this sentence along the following one:

Note that the selection of a device in a Pulser program does not enforce any choice on the backend (link to page), namely on where the simulation will be run (QPU backend vs emulator backends). Here the Device is an abstract representation that models the physics of a neutral-atom QPU, but is not the QPU itself.

File: docs/source/hardware.ipynb [hunk 7]
```diff
@@ -0,0 +1,247 @@
...
+    "\n",
+    "</div>"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Available `Devices` and where to find them"
```
Comment (ferrulli1pasqal) [inline docs/source/hardware.ipynb:55]: I'd make the title more general "How to choose a device".

Put the image at the top, and then starting the chapter like:

The choice of which device should be selected is a decision that in Pulser must be done according to the simple principle of "how close I want my physics model to be to the QPU?". 
When it comes to running a program on a QPU: Each QPU has an associated Device, which you can [get from the cloud provider you use to access this QPU](https://pulser--781.org.readthedocs.build/en/781/tutorials/backends.html#1.2.-Preparation-for-execution-on-QPUBackend).

There are several reasons for which you might want to feel less constrained by the features currently supported by a real QPU. These reasons can be to for example to test new hardware components that have not been yet installed or to run quickly design test. 
While in principle in Pulser one could create its own version of a device (i.e. its own physical model of the QPU). for user's convenience there are few device which are already avaiable for you:
- AnalogDevice. Is the closest representation to the QPU ....(basically what you have written)
-   VirtualDevice. ....

Note in the API you use the wording: _physical device_ which is not repeated here. It would be good to make it consistent.

File: docs/source/hardware.ipynb [hunk 8]
```diff
@@ -0,0 +1,247 @@
...
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "<div class=\"alert alert-info\">\n",
+    "\n",
+    "**Note**:\n",
+    "\n",
+    "It is possible to change the set of constraints with which a `Sequence` was built, by using `Sequence.switch_device`. This is especially useful if you built your `Sequence` with an example of a `Device` like `AnalogDevice`, and now want to run it on a QPU, or if you want to run it on a different QPU than initially planned.\n",
```
Comment (ferrulli1pasqal) [inline docs/source/hardware.ipynb:77]: This sounds like more a tip tbh. I would not put it here, maybe down in the page as it si a more practical info rather than a general one.

File: docs/source/hardware.ipynb [hunk 9]
```diff
@@ -0,0 +1,247 @@
...
+    "<center>\n",
+    "<img src=\"files/decision_diagram_device.png\" alt=\"Decision Diagram to select a Device for the computation\" width=\"600\">\n",
+    "</center>\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Selecting your `Device`"
```
Comment (ferrulli1pasqal) [inline docs/source/hardware.ipynb:95]: I'd make the title  "How to read the Device specifications".

File: docs/source/hardware.ipynb [hunk 10]
```diff
@@ -0,0 +1,247 @@
...
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Now that you know where to look for `Devices`, you might want know how to choose the one matching your usecase. The device specifications are here to guide your choice:"
+   ]
+  },
+  {
+   "cell_type": "code",
```
Comment (ferrulli1pasqal) [inline docs/source/hardware.ipynb:106]: I would bring this  below in the example section dedicated to the AnalogDevice (which btw I would create after the channel and make it contain channel examples as well - see later comments)

Here I would make more a list, like the one we have actually in the API https://pulser--781.org.readthedocs.build/en/781/apidoc/core.html#physical-devices and explain how to read each field. (You can reuse much of the info you have written in the bullet point list which is below as intro for each part, splitting register and layout though) For example we print Rydberg Level and Ising interaction coeff but we don't explain how they should be read. What is 60? 

This is an opportunity also to decide to not go into details for some params, so to drive the attention to the user to the params that are more important.

File: docs/source/hardware.ipynb [hunk 11]
```diff
@@ -0,0 +1,247 @@
...
+    "\n",
+    "- \"Register\" and \"Layout\" parameters are going to constrain the creation of your `Register`, that is, the number of atoms you can use and how you can place them in place. [As a reminder](programming.md#create-the-register), this impacts the [interaction strength in the interaction Hamiltonian](programming.md#interaction-hamiltonian). The creation of a `Register` is presented here. If the `Device` requires a `Layout` (as in the example here), then you have to associate a `RegisterLayout` to the `Register`, which adds more constraints. Check [this tutorial](./tutorials/reg_layouts.nblink) to see how to do it.\n",
+    "\n",
+    "- Among the \"Device\" parameters, the \"Rydberg level\" is going to complete the determination of the [interaction strength in the interaction Hamiltonian](programming.md#interaction-hamiltonian). This level determines the \"Ising interaction coefficient\", which is the $C_6$ coefficient of the [Ising Hamiltonian](./programming.md#ising-hamiltonian). The quantity $\\frac{C_6}{\\hbar}$ is accessible via `pulser.AnalogDevice.interaction_coeff`.\n",
```
Comment (ferrulli1pasqal) [inline docs/source/hardware.ipynb:124]: If C_6 is the Ising interaction coefficient, I would avoid using the brackets to refer to it. I would also enforce the use of such naming in other pages so that if I read in the specs "Ising interaction coefficient" i know without doubt that it is C_6

File: docs/source/hardware.ipynb [hunk 12]
```diff
@@ -0,0 +1,247 @@
...
+    "Other parameters are being defined in the `Device`. You can see the effect of each them in the [API documentation](./apidoc/core.rst#Devices).\n",
+    "\n",
+    "</div>"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### The `AnalogDevice`"
```
Comment (ferrulli1pasqal) [inline docs/source/hardware.ipynb:150]: I would make a different section with one example: AnalogDevice and Rydberg channel and move it after the channel.
I'd put in there the code block print of device specs, and discuss the hamiltonian which is  implemented ad you do here. In there also the rydberg global spec printed. A good reason for doing so is also that the interaction Hamiltonian presented here is exactly the general one you present in the CHannel section but spcialised to |r> and |g>.

File: docs/source/hardware.ipynb [hunk 13]
```diff
@@ -0,0 +1,247 @@
...
+    "## The `Channels`"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "The third step to writing a Pulser program is [the selection of Channels among the Device](programming.md#pick-a-device).\n",
+    "\n",
+    "As a reminder, the selection of a `Channel` defines the [interaction Hamiltonian](programming.md#interaction-hamiltonian) and [the driving Hamiltonian](programming.md#driving-hamiltonian) $H_D$.\n",
```
Comment (ferrulli1pasqal) [inline docs/source/hardware.ipynb:178]: Nice :)

File: docs/source/hardware.ipynb [hunk 14]
```diff
@@ -0,0 +1,247 @@
...
+    "import pulser\n",
+    "\n",
+    "print(pulser.AnalogDevice.channels[\"rydberg_global\"])"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "The channel is characterized by:\n",
```
Comment (ferrulli1pasqal) [inline docs/source/hardware.ipynb:202]: Same comment as done for the register. We can create a subsection which is "reading the channel specs" where you stay general making a list of bullet points and focus on the points which are  more important. Can be also good to add a link tot he API

File: docs/source/hardware.ipynb (hunk 6 above)
Comment (a-corni) [reply docs/source/hardware.ipynb:46]: My main message is: THIS IS NOT WHAT SUBMITS TO THE QPU ! 🤣 
"the selection of a device in a Pulser program does not enforce any choice on the backend" -> this is another idea, I tackle it in the simple note below (this is not important).
I could rephrase the first sentence in "The Device is an abstract representation that models the physics of a neutral-atom QPU, but is not the QPU itself.", but I am not a fan of "abstract representation" (scares non-computer scientists) and "models" (scares computer scientists, feel like it's simulation related)

Comment (ferrulli1pasqal) [reply docs/source/hardware.ipynb:46]: AHahhah I understood it then ;)

File: docs/source/hardware.ipynb (hunk 10 above)
Comment (a-corni) [reply docs/source/hardware.ipynb:106]: Can you be a bit more clear regarding the "this" ? At the moment I can't understand your proposition of re-structuration :)
To me, the list is for the API, and the elements are explicit enough. I think it makes more sense to connect it to the step-by-step guide of the programming.md 
I don't understand your point about Rydberg Level and Ising interaction coeff, to me it's explained in depth in the second bullet point. (basically, the Rydberg level is irrelevant from a user perspective, it determines the ising interaction coefficient).

File: docs/source/hardware.ipynb (hunk 11 above)
Comment (a-corni) [reply docs/source/hardware.ipynb:124]: By brackets you mean $ ? We have always written this $C_6$... (in programming.md, in Register)

File: docs/source/hardware.ipynb (hunk 14 above)
Comment (a-corni) [reply docs/source/hardware.ipynb:202]: Creating a subsection indeed, but as above I am following programming.md to regroup the specifications

File: docs/source/hardware.ipynb [hunk 15]
```diff
@@ -0,0 +1,531 @@
...
+    "editable": true,
+    "slideshow": {
+     "slide_type": ""
+    },
+    "tags": []
+   },
+   "source": [
+    "The `Channel` is going to determine the computational basis used in the driving Hamiltonian, and what is the Hamiltonian each atom sees:\n",
+    "\n",
+    "- The type of the `Channel` defines the [states](conventions.md#bases) that can be addressed by the [driving Hamiltonian](programming.md#driving-hamiltonian) if this channel is picked. All the child classes of `Channel` can be found [here](./apidoc/core.rst#available-channels).\n",
```
[the remark this answers is not in the document]
Comment (a-corni) [reply docs/source/hardware.ipynb]: As always, it worked #RTDGOAT

[the remark this answers is not in the document]
Comment (a-corni) [reply docs/source/hardware.ipynb]: I think I will have more difficulties with hiding the input of the last cell, I am trying all the tags I can think of, as here https://discourse.jupyter.org/t/difficulty-editing-metadata-tags-to-hide-cell-inputs-and-outputs/14973/3, I place some hope in the source_hidden tag https://nbformat.readthedocs.io/en/latest/format_description.html#cell-metadata, but it seems a redundant issue in nbsphinx, since 2016 https://github.com/spatialaudio/nbsphinx/issues/15#issuecomment-515746378 -> https://github.com/spatialaudio/nbsphinx/pull/543

File: pulser-core/pulser/devices/_device_datacls.py [hunk 16]
```diff
@@ -746,48 +746,48 @@ class Device(BaseDevice):
     Attributes:
```
Comment (HGSilveri) [inline pulser-core/pulser/devices/_device_datacls.py:746]: Can we check out this looks with `Args:` instead of `Attributes:`? I think it might make the rendering more compact

File: docs/source/hardware.ipynb [hunk 17]
```diff
@@ -0,0 +1,531 @@
...
+    "jupyter": {
+     "source_hidden": true
+    },
+    "hide_input": true,
+    "slideshow": {
+     "slide_type": ""
+    },
+    "tags": [
+     "hide-input"
+    ]
```
Comment (HGSilveri) [inline docs/source/hardware.ipynb:480]: It looks like you have tried every reasonable thing and it still doesn't work, so let's drop this. Sorry for pushing you into this rabbit hole :disappointed:

File: docs/source/hardware.ipynb [hunk 18]
```diff
@@ -0,0 +1,531 @@
...
+    "tags": [],
+    "vscode": {
+     "languageId": "raw"
+    }
+   },
+   "source": [
+    ".. autoclass:: pulser.devices._device_datacls.Device\n",
+    "  :noindex:"
+   ]
+  },
```
Comment (HGSilveri) [inline docs/source/hardware.ipynb:220]: I think that if we don't manage to compress this we might want to toggle it, it is very big and it breaks the flow

File: docs/source/hardware.ipynb [hunk 19]
```diff
@@ -0,0 +1,531 @@
...
+    "    - `max_radial_distance`\n",
+    "    - `min_atom_distance`\n",
+    "\n",
+    "2) The `rydberg_level` determines the  [Ising interaction coefficient](./programming.md#ising-hamiltonian) $C_6$ of the Ising Hamiltonian. The quantity $\\frac{C_6}{\\hbar}$ is accessible via the `interaction_coeff` attribute of the `Device`.\n",
+    "\n",
+    "3) The `Channels` in the `channel_objects` parameter are going to determine what [Channels are available for the computation](programming.md#pick-the-channels). Knowing what states you want to use in your computation, you can first check that they are among the `Device.supported_states`, then find the bases and their associated channel that enable to use these states using [the conventions page](conventions.md#bases).\n",
+    "\n",
+    "4) The `max_sequence_duration` constrains the duration of the [Pulses you can add](programming.md#add-the-pulses), and therefore the Hamiltonian describing the system can at most be defined between 0 and this value.\n",
+    "\n",
+    "5) The `max_runs` limits the number of runs a quantum program can be executed on the QPU. See the section on `Backends` to read more about this."
```
Comment (HGSilveri) [inline docs/source/hardware.ipynb:271]: Link once more here?

**Review by HGSilveri — COMMENTED** [review]
This is looking really nice, great job @a-corni !

**Review by HGSilveri — APPROVED** [review]
LGTM! I'd just suggest redirecting from the Programming page to here like I did here: 81954ad3e02fa6dbc1de94d857df837021f16e31

**Comment by HGSilveri** [comment]
> LGTM! I'd just suggest redirecting from the Programming page to here like I did here: [81954ad](https://github.com/pasqal-io/Pulser/commit/81954ad3e02fa6dbc1de94d857df837021f16e31)

Sorry, I was not very clear. What I meant is `See Also` boxes inside the Programming page, like these: 19881709205533f997f980b6614590ecff623a47
